UbuntuDrivers/detect.py

```diff
@@ -171,67 +203,96 @@ def packages_for_modalias(apt_cache, modalias):
             for p in bus_map[alias]:
                 pkgs.add(p)
 
-    return [apt_cache[p] for p in pkgs]
+    return [apt_cache.__getitem__(p) for p in pkgs]
 
 
 packages_for_modalias.cache_maps = {}
 
 
-def _is_package_free(pkg):
-    assert pkg.candidate is not None
+def _is_package_free(apt_cache, pkg):
+    depcache = apt_pkg.DepCache(apt_cache)
+    candidate = depcache.get_candidate_ver(pkg)
+    assert candidate is not None
     # it would be better to check the actual license, as we do not have
     # the component for third-party packages; but this is the best we can do
     # at the moment
     #
     # We can assume the NVIDIA packages to be non-free
     if pkg.name.startswith('nvidia'):
         return False
-    for o in pkg.candidate.origins:
-        if o.component in ('restricted', 'multiverse'):
-            return False
+
+    records = apt_pkg.PackageRecords(apt_cache)
+    records.lookup(candidate.file_list[0])
+
+    for pfile, _ in pkg.version_list[0].file_list:
+        if not pfile.component:
+            # This is probably from the test suite
+            try:
+                component = records.__getitem__('Component')
+                return (component not in ('restricted', 'multiverse'))
+            except KeyError:
+                return False
+        else:
+            if pfile.component in ('restricted', 'multiverse'):
+                return False
     return True
 
 
-def _is_package_from_distro(pkg):
-    if pkg.candidate is None:
+def _is_package_from_distro(apt_cache, pkg):
+    depcache = apt_pkg.DepCache(apt_cache)
+    candidate = depcache.get_candidate_ver(pkg)
+    if candidate is None:
         return False
 
-    for o in pkg.candidate.origins:
-        if o.origin == 'Ubuntu':
-            return True
-    return False
+    records = apt_pkg.PackageRecords(apt_cache)
+    records.lookup(candidate.file_list[0])
+
+    try:
+        origin = records.__getitem__('Origin')
```

Comment:
Consider using `candidate.file_list_list[0][0].origin` instead of doing the record lookup. I don't know why the field is copied to Packages files in the first place in Ubuntu.